doc/lspconfig.txt

```diff
@@ -1,242 +1,628 @@
+Examples are provided for didactic purpose:
+
+- `util.root_pattern`: variadic function which takes multiple arguments, which
+  corresponding to a different root pattern against which the contents of the
+  current directory in the traversal will be matched against using
+  `vim.fn.glob`. See `:help vim.fn.glob` for details.
+>
+   root_dir = util.root_pattern('pyproject.toml', 'requirements.txt')
+<
+- `util.find_git_ancestor`: locates the first parent directory containing a
+  `.git` directory.
+>
+   root_dir = util.find_git_ancestor
 
+- `util.find_node_modules_ancestor`:  locates the first parent directory
+  containing a `node_modules` directory.
+>
+   root_dir = util.find_node_modules_ancestor
+<
 
-Please see the [wiki](https://github.com/neovim/nvim-lspconfig/wiki) for
-additional topics, including:
-* [Installing language servers automatically](https://github.com/neovim/nvim-lspconfig/wiki/Installing-language-servers-automatically)
-* [Snippets support](https://github.com/neovim/nvim-lspconfig/wiki/Snippets)
-* [Project local settings](https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings)
-* [Recommended plugins for enhanced language server features](https://github.com/neovim/nvim-lspconfig/wiki/Language-specific-plugins)
+- `util.find_package_json_ancestor`:  locates the first parent directory
+  containing a `package.json`.
+>
+   root_dir = util.find_json_ancestor
+<
 ==============================================================================
-11. Windows                                                *lspconfig-windows*
+COMPOSITIONAL ROOT DIRECTORIES                      *lspconfig-root-composition*
 
+The `root_dir` key in `config` and `setup` can hold any function of the form
+>
+  function custom_root_dir(filename, bufnr)
+  returns nil | string
+>
+This allows for rich composition of root directory patterns which is necessary
+for some project structures. A worked example for kotlin:
+
+>
+  local root_files = {
+    'settings.gradle', -- Gradle (multi-project)
+    'settings.gradle.kts', -- Gradle (multi-project)
+    'build.xml', -- Ant
+    'pom.xml', -- Maven
+  }
+
+  local fallback_root_files = {
+    'build.gradle', -- Gradle
+    'build.gradle.kts', -- Gradle
+  }
+  root_dir = function(fname)
+    local primary = util.root_pattern(unpack(root_files))(fname)
+    local fallback = util.root_pattern(unpack(fallback_root_files))(fname)
+    return primary or fallback
+  end
+<
+Browsing the source of the default configurations is recommended.
 
-In order for neovim to launch certain executables on Windows, it must append
-`.cmd` to the command name. To work around this, manually append `.cmd` to the
-entry `cmd` in a given plugin's setup{} call.
 ==============================================================================
-12. Contributions                                    *lspconfig-contributions*
+SINGLE FILE SUPPORT                              *lspconfig-single-file-support*
```

Comment:
I don't much like this section as it's mainly complaining about lspconfig's own configurations (no matter where they come from), but if it'll be removed before 0.6 anyway, I won't argue about it.

Reply:
I'm going to keep it until I merge the PR, but even then I will still have it but change the wording to explain the limitations of the protocol

Reply:
Yeah, hence my later suggestion to turn this into a section about the role of `root_directory` at the start of the section.